### ISSUE TITLE
Consider eliminating Future's dependency on ExecutorService

### ISSUE THREAD

**Issue by 404-** [vavr-io/vavr#2194]
Currently, in order to create an `io.vavr.concurrent.Future` and instance of a `java.util.concurrent.ExecutorService` is required (assuming the `DEFAULT_EXECUTOR_SERVICE` is unacceptable). Since Java's native `java.util.concurrent.CompletableFuture` simply requires a `java.util.concurrent.Executor`, shouldn't it also be sufficient for Vavr's Futures?
The problem can be clearly illustrated when attempting to leverage Vavr's Futures when developing with Akka, since the `akka.dispatch.MessageDispatcher` does *not* implement `java.util.concurrent.ExecutorService` (nor should it).

**Comment by danieldietrich** [comment]
Currently we rely on java.util.concurrent.Future returned by an ExecutorService.submit() because of **cancellation**. That is the only dependency.

I think we can get rid of ExecutorService, similar to [this](https://gist.github.com/viktorklang/5409467). I've prototyped a version but have a remaining open point - I need to get the running Thread of the Future computation in order to be able to interrupt it if necessary. Will take a look tomorrow.

**Comment by danieldietrich** [comment]
I think I have an idea - an Executor's execute method takes a Runnable. We could 'trap' it by wrapping the unit of work in an own runnable, which stores the 'current thread' (from the view-point of the runnable) somewhere.

Then we should be done (modulo synchronization boilerplate code etc.). I will check that tomorrow.

**Comment by danieldietrich** [comment]
Hey @404-,

I would be very interested in examples on how to use Vavr's Future in conjunction with Akka (as you mentioned above). Are you interested in a guest post on the Vavr blog?

Also I'm interested in your expectations regarding handling of fatal errors, like those:

```java
    static boolean isFatal(Throwable throwable) {
        return throwable instanceof InterruptedException
                || throwable instanceof LinkageError
                || throwable instanceof ThreadDeath
                || throwable instanceof VirtualMachineError;
    }
```

See also #2198

**Comment by 404-** [comment]
@danieldietrich thank you for such a fast turnaround and the feedback.

I could write something up, but I'm not sure if there's enough content. Akka team does a decent job of duplicating APIs that return a `scala.concurrent.Future` such that the return type of each duplicated API is `java.util.concurrent.CompletionStage`. The main point of entry for Vavr is the `io.vavr.concurrent.Future#toCompletableFuture` converter. E.g., `akka.pattern.PatternsCS` has a set of APIs that parallel those in the `akka.pattern` package object:
```
package com.example;

import akka.actor.AbstractActor;
import akka.dispatch.MessageDispatcher;

import java.util.concurrent.CompletionStage;
import java.util.concurrent.Executor;

import static akka.pattern.PatternsCS.pipe;

public class AkkaWithVavrFutures extends AbstractActor {
    Service service;
    MessageDispatcher dispatcher;

    void foo() {
        pipe(service.call(dispatcher), dispatcher).to(self());
    }

    interface Service {
        class Response {}

        CompletionStage<Response> call(Executor executor);
        //desired replacement:
//        io.vavr.concurrent.Future<Response> call(Executor executor);
    }
}
```
In this scenario, imagine that I'm on a team (at a private company) which vends an internal framework based on Akka. This team controls a set of interfaces (similar to the `Service` interface above, which the clients use as entry points into the framework. As long as other APIs that clients rely on are also using Vavr classes, it would be pretty straightforward and desirable to use Vavr's Future over Java's CompletionStage.
Is this the kind of content you had in mind for a blog post?

Regarding fatal errors, I'm not sure if anything else needs to be done in addition to re-throwing them - after all, there's no way for an application to recover from them. Just let it crash, right? Or am I missing something?

**Comment by danieldietrich** [comment]
@404- thanks for the detailed reply!

> I could write something up, but I'm not sure if there's enough content. (...) Is this the kind of content you had in mind for a blog post?

Yes, even a one-pager would be interesting for the audience, e.g. a hello world with Akka + Vavr.
For example showing a minimal setup with Maven or Gradle. We have a quite big audience :)

The easiest way is to prepare it as markdown, e.g. in a private Gist. I will put it all on the blog then.

> Regarding fatal errors, I'm not sure if anything else needs to be done in addition to re-throwing them (...) Just let it crash, right? Or am I missing something?

Yes, letting the application crash in the presence of fatal errors seems the way to go.

Thx!

**Comment by 404-** [comment]
@danieldietrich can you take a look at the code and let me know if you'd like me to add or change anything: https://gist.github.com/404-/6cad39b36f51c9e0136404ef16be5fc4
I'll do the actual write-up once we're happy with the code.

**Comment by danieldietrich** [comment]
@404- Many thanks, I will do that later today!

**Comment by danieldietrich** [comment]
@404- your code looks great! I like the `$InstanceOf` pattern, we should integrate it in Vavr. I will create an issue.

**Comment by 404-** [comment]
Sorry for the delay, here's the whole thing (including the blog post itself): https://gist.github.com/404-/270c335df382132ef30251b5e5b14c0b

Can you take a look and let me know if you'd like me to add or amend anything?

**Comment by danieldietrich** [comment]
Hi @404-, thank you! Could you please add a short introduction to yourself? I will add it as guest post to blog.vavr.io. The reader should know that the credits go to you (optionally with profile pic).

**Comment by 404-** [comment]
@danieldietrich good point, I've updated my GitHub profile. Were you looking for any other info or is what I provided sufficient?

By the way, when I was writing the code, I wasn't sure if it was useful to include the two helper methods (i.e., the entirety of `TemporaryWorkarounds.java`) since you were planning on addressing them in Vavr itself. Feel free to completely remove them, as if the code samples were written against an upcoming version of Vavr where something like `$InstanceOf` is already present and `Future.of` no longer requires an `ExecutorService`.
